src/GraphCoarsen.cpp

```diff
@@ -13,788 +13,1100 @@
-
-                if (build_coarse_relation)
-                {
-                    mfem::DenseMatrix& CM_el_loc1(CM_el[Agg1]);
-                    mfem::DenseMatrix& CM_el_loc2(CM_el[Agg2]);
-
-                    id2_in_Agg1 = edge_cdof_marker[col];
-                    if (Aggs.Size() == 1)
-                    {
-                        CM_el_loc1(id1_in_Agg1, id2_in_Agg1) += entry_value;
-                        CM_el_loc1(id2_in_Agg1, id1_in_Agg1) += entry_value;
-                    }
-                    else
-                    {
-                        assert(Aggs.Size() == 2);
-                        CM_el_loc1(id1_in_Agg1, id2_in_Agg1) += entry_value / 2.;
-                        CM_el_loc1(id2_in_Agg1, id1_in_Agg1) += entry_value / 2.;
-                        id2_in_Agg2 = edge_cdof_marker2[col];
-                        CM_el_loc2(id1_in_Agg2, id2_in_Agg2) += entry_value / 2.;
-                        CM_el_loc2(id2_in_Agg2, id1_in_Agg2) += entry_value / 2.;
-                    }
-                }
-                else
-                {
-                    CoarseM_->Add(row, col, entry_value);
-                    CoarseM_->Add(col, row, entry_value);
-                }
-            }
+        int num_bubbles_i = vertex_targets_[agg].Cols() - 1;
+
+        for (int j = 0; j < num_bubbles_i; ++j)
+        {
+            D_coarse.Add(counter + agg + 1 + j, total_traces + counter + j, 1.0);
         }
+
+        counter += num_bubbles_i;
     }
-    mfem::SparseMatrix newPedges(Pedges_i, Pedges_j, Pedges_data,
-                                 nedges, total_num_traces + bubble_counter);
-    Pedges.Swap(newPedges);
 
-    if (!build_coarse_relation)
-        CoarseM_->Finalize(0);
+    return D_coarse.ToSparse();
 }
 
-void GraphCoarsen::BuildW(const mfem::SparseMatrix& Pvertices)
+/// Helper function to build fine-level aggregate sub-M
+/// corresponding to dofs on a face
+void BuildAggregateFaceM(const MixedMatrix& mgl, int agg,
+                         const std::vector<int>& edge_dofs_on_face,
+                         const SparseMatrix& vertex_agg,
+                         const SparseMatrix& edge_vertex,
+                         std::vector<int>& col_marker,
+                         Vector& M_local)
 {
-    if (W_proc_)
+    const auto& M_fine_elem = mgl.GetElemM();
+    const auto& partition = vertex_agg.GetIndices();
+
+    int fine_size = edge_dofs_on_face.size();
+    M_local.SetSize(fine_size);
+
+    for (int i = 0; i < fine_size; ++i)
     {
-        unique_ptr<mfem::SparseMatrix> W_tmp(mfem::RAP(Pvertices, *W_proc_, Pvertices));
-        mfem::SparseMatrix W_thresh(*W_tmp);
+        int edge_dof = edge_dofs_on_face[i];
+        auto verts = edge_vertex.GetIndices(edge_dof);
+        int vert = (partition[verts[0]] == agg ) ? verts[0] : verts[1];
+
+        std::vector<int> fine_edges = mgl.GetElemDof().GetIndices(vert);
+        SetMarker(col_marker, fine_edges);
```

Comment:
It seems to me that complexity of the original implementation (using `break`) is lower, isn't it?